tikv/pd
# etcdutil, leadership: make more high availability

### What problem does this PR solve?


Issue Number: Close #6554 https://github.com/tikv/pd/issues/6442

### What is changed and how does it work?

#### leadership

1. make leadership exit watch loop as soon as possible when etcd is unavailable

2. add some other logic similar to loopwatcher, such as `revision++`, `wresp.Err() != nil` and `WithRequireLeader`

#### etcd client

Because the etcd client's keepalive is based on the grpc connection, it can only determine if the network is working properly.

And because etcd client's balancer only supports round-robin policy, it cannot be set to connect only to the leader.

So when the etcd client's original connection is not working for various reasons, such as network isolation, leader-io-hang, or tcp half-open connection, we need to switch the connection to an available node in a more timely manner.

It is easy to think that we can create additional goroutines to check the health of the connections, so for each etcd client, we will create two goroutines, one goroutine to check the health of the node and another goroutine to synchronize the members.

And grpc substream retry policy is exponential rather than linear, which also needs to change it.

```commit-message
```

### Check List



Tests



- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
1. pd-leader-io-hang

   ![image](https://github.com/pingcap/tidb/assets/19542290/569bf41c-4ce0-4723-832b-6d128bc32f65)
   only update pd etcd client, there are two 8027 error
   ![image](https://github.com/pingcap/tidb/assets/19542290/0957e457-11e1-42b9-873b-d5fda258f52a)
   update pd and tidb etcd client at the same, there is no any 8027 error


2. pd-leader-to-all-network-partition

    ![image](https://github.com/pingcap/tidb/assets/19542290/8fa5998f-4588-4d77-af59-5261ab087f92)
    only update pd etcd client, there is a warning with 21.7% drop.
    ![image](https://github.com/pingcap/tidb/assets/19542290/297ff2bf-b9b5-47d0-9f80-9ab94eff03ca)
    update pd and tidb etcd client at the same, there are some warnings, but all warnings are just a little over 20% and below 25%

3. Manual stop and random restart api server, observe when tso server watch primary
https://github.com/tikv/pd/pull/6577#issuecomment-1597015549
![image](https://github.com/tikv/pd/assets/19542290/06c9ade3-2d5c-4b3a-a973-401f512d2fe9)
![image](https://github.com/tikv/pd/assets/19542290/d6971fa3-3424-4f9d-b2d5-0539b34bac4b)

4. Test 30 times

    Step: `kubectl delete pod serverless-cluster-pd-0 serverless-cluster-pd-1 serverless-cluster-pd-2 -n tidb-serverless --force --grace-period=0`
    note: replace tso server with https://github.com/tikv/pd/pull/6577 use pd-tso-bench https://github.com/tikv/pd/pull/6699
    Result: can recovery in 2 min




### Release note



```release-note
None.
```
